File: muxer-server/src/main/java/pt/um/tf/taskmux/server/TaskQueues.java
package pt.um.tf.taskmux.server;

import pt.um.tf.taskmux.commons.IndexedDeque;
import pt.um.tf.taskmux.commons.error.DuplicateException;
import pt.um.tf.taskmux.commons.error.NoAssignableTasksException;
import pt.um.tf.taskmux.commons.task.Result;
import pt.um.tf.taskmux.commons.task.Task;

import java.net.URI;
import java.util.*;
import java.util.stream.IntStream;

public class TaskQueues {
    private final IndexedDeque<Task> inbound;
    private final Map<String, Map<URI, Task>> outbound;
    private final Map<String, Iterator<Map.Entry<String, Map<URI, Task>>>> iteratorMap;

    public TaskQueues() {
        inbound = new IndexedDeque<>();
        outbound = new HashMap<>();
        iteratorMap = new HashMap<>();
    }


    public void addToFrontInbound(Task task) {
        inbound.addFirst(task);
    }

    public void addAllToBackInbound(Collection<Task> tasks) {
        tasks.forEach(inbound::addLast);
    }

    public void clearTasks(String toClear) {
        if (outbound.containsKey(toClear)) {
            outbound.get(toClear).clear();
            outbound.remove(toClear);
        }
    }

    public void replaceAfterCompleted(String sender, Collection<Task> tasks) {
        var map = outbound.get(sender);
        outbound.put(sender, tasks.stream()
                                  .collect(HashMap::new,
                                           (t, r) -> t.put(r.getURI(), r),
                                           HashMap::putAll));
    }

    public void replaceAfterGet(Collection<Task> in, String sender, int count) {
        Map<URI, Task> nMap = in.stream()
                                .collect(HashMap::new,
                                         (h, t) -> h.put(t.getURI(), t),
                                         HashMap::putAll);
        outbound.put(sender, nMap);
        IntStream.range(0, count)
                 .forEach(i -> inbound.removeFirst());
    }

    public Result<Task> sendToOutbound(String sender) {
        Result<Task> taskResult;
        if (!outbound.containsKey(sender)) {
            outbound.put(sender, new HashMap<>());
        }
        if(inboundIsEmpty()) {
            taskResult = new TaskResult(false, null, new NoAssignableTasksException());
        }
        else {
            var in = inbound.removeFirst();
            var out = outbound.get(sender);
            if (!out.containsKey(in.getURI())) {
                out.put(in.getURI(), in);
                taskResult = new TaskResult(true, in, null);
            }
            else {
                taskResult = new TaskResult(false, null, new DuplicateException());
            }
        }
        return taskResult;
    }

    public Optional<ArrayList<Task>> backToInbound(String mem) {
        Optional<ArrayList<Task>> backToInbound = Optional.empty();
        if (outbound.containsKey(mem)) {
            var out = outbound.get(mem);
            if(!out.isEmpty()) {
                var newArr = new ArrayList<Task>(out.size());
                out.forEach((a, task) -> {
                    inbound.addFirst(task);
                    newArr.add(task);
                });
                backToInbound = Optional.of(newArr);
                out.clear();
            }
            outbound.remove(mem);
        }
        return backToInbound;
    }

    public void sendToInbound(Task task) {
        inbound.addLast(task);
    }

    public boolean AreBothEmpty() {
        return inbound.isEmpty() && outbound.isEmpty();
    }

    public boolean inboundIsEmpty() {
        return inbound.isEmpty();
    }

    public boolean outboundIsEmpty() {
        return outbound.isEmpty();
    }

    public Iterator<Map.Entry<String, Map<URI, Task>>> getOutboundIterator(String receiver) {
        Iterator<Map.Entry<String, Map<URI, Task>>> it;
        if(!iteratorMap.containsKey(receiver)) {
            it = outbound.entrySet().iterator();
            iteratorMap.put(receiver, it);
        }
        else {
            it = iteratorMap.get(receiver);
        }
        return it;
    }

    public Optional<Collection<Task>> purgeOutbound(String sender, URI url) {
        Optional<Collection<Task>> tasks = Optional.empty();
        if (outbound.containsKey(sender)) {
            var out = outbound.get(sender);
            if (out.remove(url) != null) {
                tasks = Optional.of(new ArrayList<>(outbound.get(sender).values()));
            }
        }
        return tasks;
    }

    public Collection<Task> getInbound() {
        return new ArrayList<>(inbound);
    }

    public Optional<Collection<Task>> getOutbound(String sender) {
        Optional<Collection<Task>> t = Optional.empty();
        if(outbound.containsKey(sender)) {
            t = Optional.of(new ArrayList<>(outbound.get(sender).values()));
        }
        return t;
    }

    public void wipe() {
        inbound.clear();
        outbound.clear();
    }

    public void purgeIterator(String receiver) {
        iteratorMap.remove(receiver);
    }

}

File: muxer-server/src/main/kotlin/pt/um/tf/taskmux/server/TaskQueues.kt
package pt.um.tf.taskmux.server

import pt.um.tf.taskmux.commons.IndexedDeque
import pt.um.tf.taskmux.commons.error.DuplicateException
import pt.um.tf.taskmux.commons.error.NoAssignableTasksException
import pt.um.tf.taskmux.commons.task.Result
import pt.um.tf.taskmux.commons.task.Task
import pt.um.tf.taskmux.server.messaging.OutboundMessage
import java.net.URI


class TaskQueues {
    private val inbound : IndexedDeque<Task<out Any>> = IndexedDeque()
    private val outbound : MutableMap<String, MutableMap<URI, Task<out Any>>> = mutableMapOf()
    private val iteratorMap : MutableMap<String, Iterator<Map.Entry<String, MutableMap<URI, Task<out Any>>>>> = mutableMapOf()

    fun addToFrontInbound(task : Task<out Any>) {
        inbound.addFirst(task)
    }

    fun addAllToBackInbound(tasks : Set<Task<out Any>>) {
        tasks.forEach(inbound::addLast)
    }

    fun clearTasks(toClear : String) {
        outbound[toClear]?.clear()
        outbound.remove(toClear)
    }

    fun replaceAfterCompleted(sender : String,
                              tasks : Set<Task<out Any>>) {
        outbound[sender] = tasks.associateByTo(mutableMapOf(),
                                               Task<out Any>::uri)
    }

    fun replaceAfterGet(inSet : Set<Task<out Any>>, sender : String, count : Int) {
        outbound[sender] = inSet.associateByTo(mutableMapOf(), Task<out Any>::uri)
        repeat(if (count > inbound.size) inbound.size else count) {
            inbound.removeFirst()
        }
    }

    fun sendToOutbound(sender : String) : Result<Task<out Any>> {
        if (!outbound.containsKey(sender)) {
            outbound[sender] = mutableMapOf()
        }
        return if(inboundIsEmpty()) {
            TaskResult(false, null, NoAssignableTasksException())
        }
        else {
            val inSet = inbound.removeFirst()
            val out = outbound[sender]
            out?.put(inSet.uri, inSet)
            if (out != null) {
                TaskResult(true, inSet, null)
            } else {
                //Should never happen
                TaskResult(false, null, DuplicateException())
            }
        }
    }

    fun backToInbound(mem : String) : Set<Task<out Any>> {
        val out = outbound[mem]
        val backToInbound = setOf(*out?.values?.toTypedArray() ?: arrayOf())
        backToInbound.forEach {
            inbound.addFirst(it)
        }
        outbound.remove(mem)
        return backToInbound
    }

    fun sendToInbound(task : Task<out Any>) = inbound.addLast(task)

    fun AreBothEmpty() : Boolean = inbound.isEmpty() && outbound.isEmpty()

    fun inboundIsEmpty() : Boolean = inbound.isEmpty()

    fun outboundIsEmpty() : Boolean = outbound.isEmpty()

    fun getOutboundIterator(receiver : String) : Iterator<Map.Entry<String, Map<URI, Task<out Any>>>>? {
        return if(!iteratorMap.containsKey(receiver)) {
            val it = outbound.entries.iterator()
            iteratorMap[receiver] = it
            it
        }
        else {
            iteratorMap[receiver]
        }
    }

    fun purgeOutbound(sender : String, url : URI) : Set<Task<out Any>> {
        return if (outbound.containsKey(sender)) {
            val out = outbound[sender]
            if (out?.remove(url) != null) {
                setOf(*out.values.toTypedArray())
            } else {
              emptySet()
            }
        } else {
          emptySet()
        }
    }

    fun getInbound() : Set<Task<out Any>> = inbound.toSet()

    fun getOutbound(sender : String) : Set<Task<out Any>> {
        return outbound[sender]?.values?.toSet() ?: emptySet()
    }

    fun wipe() {
        inbound.clear()
        outbound.clear()
    }

    fun purgeIterator(receiver: String) {
        iteratorMap.remove(receiver)
    }
}
